Pairwise sequence alignment refers to the process of comparing two nucleotide or amino acid sequences to identify regions of similarity. There are two broad types of alignment:



*   Global alignment aligns both sequences from end to end, aligning each
character in each sequence only once.
*   Local alignment aligns subsequences of each sequence, rather than the entire sequence.

Pairwise sequence alignment can be performed in Python using the Biopython suite of tools. Here, the `Bio.Align` package will be used to perform local and global alignments. The `pairwise2` package could also be used for the purpose, but it has been deprecated. Two amino acid sequences sourced from the NCBI database - the first subunit of the COX1 protein found in *Saccharomyces cerevisiae* and *Candida albicans* respectively - will be aligned.

The effects of applying penalties for mismatches or gaps in the two sequences will also be explored.



In [1]:
!pip install biopython

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 41.7 MB/s eta 0:00:00


In [34]:
from Bio import Align
from Bio.Align import substitution_matrices
aligner = Align.PairwiseAligner()

In [45]:
cox1_s_cerevisiae = "MVQRWLYSTNAKDIAVLYFMLAIFSGMAGTAMSLIIRLELAAPGSQYLHGNSQLFNVLVVGHAVLMIFFLVMPALIGGFGNYLLPLMIGATDTAFPRINNIAFWVLPMGLVCLVTSTLVESGAGTGWTVYPPLSSIQAHSGPSVDLAIFALHLTSISSLLGAINFIVTTLNMRTNGMTMHKLPLFVWSIFITAFLLLLSLPVLSAGITMLLLDRNFNTSFFEVSGGGDPILYEHLFWFFGHPEVYILIIPGFGIISHVVSTYSKKPVFGEISMVYAMASIGLLGFLVWSHHMYIVGLDADTRAYFTSATMIIAIPTGIKIFSWLATIHGGSIRLATPMLYAIAFLFLFTMGGLTGVALANASLDVAFHDTYYVVGHFHYVLSMGAIFSLFAGYYYWSPQILGLNYNEKLAQIQFWLIFIGANVIFFPMHFLGINGMPRRIPDYPDAFAGWNYVASIGSFIATLSLFLFIYILYDQLVNGLNNKVNNKSVIYNKAPDFVESNTIFNLNTVKSSSIEFLLTSPPAVHSFNTPAVQS"

cox1_c_albicans = "MSYVTRWLFSTSHKDIGILYLIYGMVSAMVATGMSVIIRLELSGPGPMFLHGNNQVFNVLVTGHAIAMIFLFVMPILIGSFGNYFLPIMIGAVDMAFARLNNISFWCLPPALVCVIASVLIETGAGTGWTVYPPLSSISAHSGPSVDLAIFAIHLTSISSLLGAINFIVTTLNMRSIGIHMIDMPLFVWAIFFTAILLLLSLPVLTAGVTLLLMDRNFNTGFYEVAAGGDPILYEHLFYFFGHPEVYIIIIPGFGIISHVISTYTKKPIFGQIGMIYAIGSIGLLGFLVWSHHMYVVGLDIDSRAYFTSATMVIAIPTGIKIFSWLATLYGGELRLAVPMLFALGFLFLFTIGGLTGVMLSNASIDVAFHDTYYVVGHFHYVLSMGALFSLIAGYYYWGPAMFGLKYNKVLAEVHYWLLFVSVNIIFLPMHFLGLNGMPRRIPQYPDAFVGWNVVSSWGSIMSVISVLIGLYSVLVQLTNGENEREEIQVTPDYLESNNTRDVRDSDLELITSRPAQYHTFSELPILIQQI"

First, the two sequences will be aligned globally, with default parameters. There are no gap penalties and identical characters are scored 1, else 0.

In [46]:
global_alignments = aligner.align(cox1_s_cerevisiae, cox1_c_albicans)
print(global_alignments[0])

target            0 M--VQ-RWLY-STNA--KDIAV--LYFMLAIFS-GM--AGT--A--MSL-IIRLELAA--
                  0 |--|--|||--||----|||----||--|-|---||--|----|--||--||||||----
query             0 MSYV-TRWL-FST--SHKDI--GILY--L-I--YGMVSA--MVATGMS-VIIRLEL--SG

target           42 PGSQY---LHGNS-QL-FNVLVV-GHAVL--MIFFL-VMPA-LIGG-FGNY-LLPL-MIG
                 60 ||------||||--|--|||||--|||----|||-|-|||--|||--||||-|-|--|||
query            44 PG---PMFLHGN-NQ-VFNVLV-TGHA--IAMIF-LFVMP-ILIG-SFGNYFL-P-IMIG

target           89 AT-DT-AFP-RI-NNIA-FWV-LPMG--LVCLVT--STLV--ES-GAGTGWTVYPPLSSI
                120 |--|--||--|--|||--||--||----|||-|---|--|--|--|||||||||||||||
query            91 A-VD-MAF-AR-LNNI-SFW-CLP--PALVC-V-IAS--VLIE-TGAGTGWTVYPPLSSI

target          136 Q-AHSGPSVDLAIFAL-HLTSISSLLGAINFIVTTLNMRTN--G--MT--MHKLPLFVWS
                180 --|||||||||||||--||||||||||||||||||||||----|--|---|---|||||-
query           138 -SAHSGPSVDLAIFA-IHLTSISSLLGAINFIVTTLNMR--SIGIHM-IDM---PLFVW-

target          188 -IFI

In [47]:
#The following code returns the indices of the subsequences that aligned to each other in the first alignment.
global_alignments[0].aligned

array([[[  0,   1],
        [  1,   2],
        [  3,   6],
        [  7,   9],
        [ 11,  14],
        [ 16,  18],
        [ 20,  21],
        [ 22,  23],
        [ 25,  27],
        [ 27,  28],
        [ 30,  31],
        [ 31,  33],
        [ 34,  40],
        [ 42,  44],
        [ 47,  51],
        [ 52,  53],
        [ 54,  59],
        [ 60,  63],
        [ 65,  68],
        [ 69,  70],
        [ 70,  73],
        [ 74,  77],
        [ 78,  82],
        [ 82,  83],
        [ 84,  85],
        [ 86,  90],
        [ 91,  92],
        [ 93,  95],
        [ 96,  97],
        [ 98, 101],
        [102, 104],
        [105, 107],
        [109, 112],
        [113, 114],
        [115, 116],
        [118, 119],
        [119, 120],
        [121, 136],
        [137, 150],
        [151, 173],
        [175, 176],
        [176, 177],
        [178, 179],
        [182, 187],
        [188, 190],
        [191, 193],
        [194, 203],
        [204, 206],
        [207, 208],
        [209, 212],


In [48]:
# The score for an alignment can be accessed with the `score` attribute

global_alignments[0].score

379.0

We see that the first alignment has a score of 379. As we are only applying a score of 1 to all matches, and one of 0 to gaps and mismatches, it follows that, in the first alignment, the two sequences have 379 bases in common.

In [52]:
print(len(cox1_s_cerevisiae))
print(len(cox1_c_albicans))

534
531


In [53]:
total_residues = len(cox1_s_cerevisiae) + len(cox1_c_albicans)
identical_prop = global_alignments[0].score / total_residues
print(identical_prop)

0.355868544600939


From the given calculation, we can estimate that approximately 35% of the total compared amino acids are identical.

We will now see what happens when mismatches are penalised. For each mismatch, 1 points will be deducted from the total score. Matches will be awarded 2 points. Gaps will not be penalised.

In [54]:
aligner.match_score = 2
aligner.mismatch_score = -1
penalised_global_alignments = aligner.align(cox1_s_cerevisiae, cox1_c_albicans)
print(penalised_global_alignments[0])

target            0 M--VQ-RWLY-STNA--KDIAV--LYFMLAIFS-GM--AGT--A--MSL-IIRLELAA--
                  0 |--|--|||--||----|||----||--|-|---||--|----|--||--||||||----
query             0 MSYV-TRWL-FST--SHKDI--GILY--L-I--YGMVSA--MVATGMS-VIIRLEL--SG

target           42 PGSQY---LHGNS-QL-FNVLVV-GHAVL--MIFFL-VMPA-LIGG-FGNY-LLPL-MIG
                 60 ||------||||--|--|||||--|||----|||-|-|||--|||--||||-|-|--|||
query            44 PG---PMFLHGN-NQ-VFNVLV-TGHA--IAMIF-LFVMP-ILIG-SFGNYFL-P-IMIG

target           89 AT-DT-AFP-RI-NNIA-FWV-LPMG--LVCLVT--STLV--ES-GAGTGWTVYPPLSSI
                120 |--|--||--|--|||--||--||----|||-|---|--|--|--|||||||||||||||
query            91 A-VD-MAF-AR-LNNI-SFW-CLP--PALVC-V-IAS--VLIE-TGAGTGWTVYPPLSSI

target          136 Q-AHSGPSVDLAIFAL-HLTSISSLLGAINFIVTTLNMRTN--G--MT--MHKLPLFVWS
                180 --|||||||||||||--||||||||||||||||||||||----|--|---|---|||||-
query           138 -SAHSGPSVDLAIFA-IHLTSISSLLGAINFIVTTLNMR--SIGIHM-IDM---PLFVW-

target          188 -IFI

In [55]:
penalised_global_alignments[0].score

758.0

In [56]:
print(global_alignments[0].score)

#Theoretically, we assume that this score means that all the characters matched in the first alignment, and there were no gaps or mismatches.
#If, as above, we award matches with +2, then:

ideal_score = 2 * global_alignments[0].score
print(ideal_score)

#We see that this exactly matches the score that we obtained above, while scoring matches with 2, penalising mismatches by 1 and scoring gaps as 0.
#We may conclude, from this calculation, that there are no mismatches in the first global alignment. However, there may be gaps.
#Printing out the first alignment, as we have done above, confirms this brief analysis.


379.0
758.0


Now, in addition to the mismatch penalty specified previously, we will penalise gaps.

0.5 points will be deducted for opening a gap.

0.1 points will be deducted for extending it.

In [57]:
aligner.match_score = 2
aligner.mismatch_score = -1
aligner.open_gap_score = -0.5
aligner.extend_gap_score = -0.1
aligner.target_end_gap_score = 0.0
aligner.query_end_gap_score = 0.0

gap_penalised_global_alignments = aligner.align(cox1_s_cerevisiae, cox1_c_albicans)
print(gap_penalised_global_alignments[0])

target            0 M--VQRWLYST--NAKD--IAVLYFMLAI-FSGM-----A-GTAMSLIIRLEL--AAPG-
                  0 |--|.|||.||----||--|--||--|-|---||-----|-|--||.||||||----||-
query             0 MSYVTRWLFSTSH--KDIGI--LY--L-IY--GMVSAMVATG--MSVIIRLELSG--PGP

target           44 --SQYLHGNSQLFNVLVVGH--AVLMIF-FLVMPALIGGFGNYLLPLMIGATDTAFPRIN
                 60 -----||||.|.|||||.||--|--|||-|-|||.|||.||||.||.||||.|.||.|.|
query            47 MF---LHGNNQVFNVLVTGHAIA--MIFLF-VMPILIGSFGNYFLPIMIGAVDMAFARLN

target           99 NIAFWVLP--MGLVCLV--TSTLV--ESGAGTGWTVYPPLSSIQAHSGPSVDLAIFALHL
                120 ||.||.||----|||-|---|--|--|.|||||||||||||||.|||||||||||||.||
query           101 NISFWCLPPA--LVC-VIA-S--VLIETGAGTGWTVYPPLSSISAHSGPSVDLAIFAIHL

target          153 TSISSLLGAINFIVTTLNMR--TNG--M--TMHKLPLFVWSIFITAFLLLLSLPVLSAGI
                180 ||||||||||||||||||||----|--|---|---|||||.||.||.|||||||||.||.
query           155 TSISSLLGAINFIVTTLNMRSI--GIHMID-M---PLFVWAIFFTAILLLLSLPVLTAGV

target          207 TMLL

In [58]:
round(gap_penalised_global_alignments[0].score, 1)

649.1

As expected, the penalisation of gaps has resulted in a decrease in the final score.

Next, the sequences will be aligned locally, with default parameters. There are no gap penalties and identical characters are scored 1, else 0.

In [59]:
aligner.mode = 'local'

#Restoring all the score attributes to their default values
aligner.match_score = 1
aligner.mismatch_score = 0
aligner.open_gap_score = 0.0
aligner.extend_gap_score = 0.0
aligner.target_end_gap_score = 0.0
aligner.query_end_gap_score = 0.0

local_alignments = aligner.align(cox1_s_cerevisiae, cox1_c_albicans)
print(local_alignments[0])

target            0 M--VQ-RWLY-STNA--KDIAV--LYFMLAIFS-GM--AGT--A--MSL-IIRLELAA--
                  0 |--|--|||--||----|||----||--|-|---||--|----|--||--||||||----
query             0 MSYV-TRWL-FST--SHKDI--GILY--L-I--YGMVSA--MVATGMS-VIIRLEL--SG

target           42 PGSQY---LHGNS-QL-FNVLVV-GHAVL--MIFFL-VMPA-LIGG-FGNY-LLPL-MIG
                 60 ||------||||--|--|||||--|||----|||-|-|||--|||--||||-|-|--|||
query            44 PG---PMFLHGN-NQ-VFNVLV-TGHA--IAMIF-LFVMP-ILIG-SFGNYFL-P-IMIG

target           89 AT-DT-AFP-RI-NNIA-FWV-LPMG--LVCLVT--STLV--ES-GAGTGWTVYPPLSSI
                120 |--|--||--|--|||--||--||----|||-|---|--|--|--|||||||||||||||
query            91 A-VD-MAF-AR-LNNI-SFW-CLP--PALVC-V-IAS--VLIE-TGAGTGWTVYPPLSSI

target          136 Q-AHSGPSVDLAIFAL-HLTSISSLLGAINFIVTTLNMRTN--G--MT--MHKLPLFVWS
                180 --|||||||||||||--||||||||||||||||||||||----|--|---|---|||||-
query           138 -SAHSGPSVDLAIFA-IHLTSISSLLGAINFIVTTLNMR--SIGIHM-IDM---PLFVW-

target          188 -IFI

In [60]:
local_alignments[0].score

379.0

We can also use known substitution matrices, such as the BLOSUM62 matrix, instead of specifying gap and mismatch penalties. Incidentally, the BLOSUM62 matrix is also used by BLAST, a popular and powerful tool used to perform pairwise sequence alignments.

In [61]:
aligner_sub = Align.PairwiseAligner()
aligner_sub.substitution_matrix = substitution_matrices.load("BLOSUM62")
blosum_alignments = aligner_sub.align(cox1_s_cerevisiae, cox1_c_albicans)
print(blosum_alignments[0])

target            0 M--VQ-RWLYSTNA-KDIA-VLYFMLAIFSGM--A----GTAMSLIIRLELAA-PGSQ--
                  0 |--|--|||.||.--|||--.||--|-|.-||--|----|--||.||||||.--||----
query             0 MSYV-TRWLFSTS-HKDI-GILY--L-IY-GMVSAMVATG--MSVIIRLELS-GPG--PM

target           46 YLHGNSQLFNVLVV-GHAVL-MIF-FLVMPA-LIGG-FGNY-LLPLMIGAT-DT-AFP-R
                 60 .||||.|.|||||--|||.--|||-|-|||--|||--||||-|-|.||||--|--||--|
query            48 FLHGNNQVFNVLV-TGHAI-AMIFLF-VMP-ILIG-SFGNYFL-PIMIGA-VD-MAF-AR

target           97 INNIAFWV-LPMG--LVCLVT--ST-LVESGAGTGWTVYPPLSSIQ-AHSGPSVDLAIFA
                120 .|||.||--||----|||-|---|--|.|.|||||||||||||||--|||||||||||||
query            99 LNNISFW-CLP--PALVC-V-IAS-VLIETGAGTGWTVYPPLSSI-SAHSGPSVDLAIFA

target          150 LHLTSISSLLGAINFIVTTLNMRTN-GMTMHKL---PLFVWSIFI-TAF-LLLLSLPVLS
                180 .||||||||||||||||||||||.--|.--|-.---|||||.||--||--|||||||||.
query           152 IHLTSISSLLGAINFIVTTLNMRS-IGI--H-MIDMPLFVWAIF-FTA-ILLLLSLPVLT

target          204 AGIT

In [62]:
blosum_alignments[0].score

2153.0

The two sequences were sourced from the [NCBI database](https://www.ncbi.nlm.nih.gov/).

The documentation for the packages used was obtained from the [Biopython website](https://biopython.org/docs/1.76/api/Bio.Align.html).